src/helper/UserSession.cpp

```diff
@@ -224,6 +248,7 @@ namespace SDDM {
             qCritical() << "setgid(" << pw.pw_gid << ") failed for user: " << username;
             exit(Auth::HELPER_OTHER_ERROR);
         }
+        qputenv("XDG_RUNTIME_DIR", QByteArrayLiteral("/run/user/") + QByteArray::number(pw.pw_uid));
```

Comment:
why?

Reply:
For WaylandSocketWatcher to work.

Reply:
This looks not great. Either something like `pam_systemd` creates it and it can be used or sddm has to create its own and use that path. Just guessing that it will be `/run/user/<uid>` smells like a hack.

Reply:
pam_systemd does create it

Reply:
Exactly, but then it should only be used in the context that `$XDG_RUNTIME_DIR` is actually set in

Reply:
In FreeBSD for instance (or with ConsoleKit in general, not sure), `XDG_RUNTIME_DIR` is actually set up by the session scripts and somewhere below `/tmp` IIRC.

Reply:
On FreeBSD both `consolekit2` ([upstream](https://github.com/ConsoleKit2/ConsoleKit2/blob/1.2.4/configure.ac#L160-L167) + [downstream](https://github.com/freebsd/freebsd-ports/blob/594ba3e582e7/sysutils/consolekit2/Makefile#L53)) and `pam_xdg` ([downstream](https://github.com/freebsd/freebsd-ports/blob/594ba3e582e7/sysutils/pam_xdg/Makefile#L23)) use `/var/run/user/<uid>`, hardcoded at build-time.

Reply:
I mean, `XDG_RUNTIME_DIR` set to `/tmp` (as opposed to fallback when unset) shouldn't be used outside of debugging as [basedir-spec](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) says "The directory MUST be owned by the user, and he MUST be the only one having read and write access to it. Its Unix access mode MUST be 0700."

Reply:
Right. If something else (e.g. pam_systemd) created XDG_RUNTIME_DIR, then you must READ the environment variable XDG_RUNTIME_DIR to find out its value. Or maybe there is a d-bus interface to ask for it if this is not your current session, I don't know. It would be incorrect to guess what the value might be and set XDG_RUNTIME_DIR to that guessed value.